[Feature] 내 핏프티 수정 UI, 앨범 목록에서 사진 가져오기 구현

dudtjs1021ej opened the pull request:
### 📕 Issue

>
### 📙 작업 내역

- [x] 기존에 `UploadCodyViewController` 네이밍을 `MyFitftyViewController` 로 변경
- [x]  MyFitftyType(`modifyMyFitfty`, `uploadMyFitfty` )에 따라 내 핏프티 수정, 등록 VC 생성
- [x] 내핏프티 등록, 수정 글자 제한 구현 (2200자)
- [x] 앨범 목록 UI & 로직 구현


### 📘 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트


### 📋 체크리스트

- [x] Merge 하는 브랜치가 올바른가?
- [x] PR과 관련없는 변경사항이 없는가?
- [ ] 내 코드에 대한 자기 검토가 되었는가?
- [ ] 변경사항이 효과적이거나 동작이 작동한다는 것을 보증하는 테스트를 추가하였는가?
- [ ] 새로운 테스트와 기존의 테스트가 변경사항에 대해 만족하는가?

### 📱 테스트 방법



### 📝 리뷰 노트

> 구현 시에 고민이었던 점들 혹은 특정 부분에 대한 의도가 있다면 PR 리뷰를 돕기 위해 서술합니다. 리뷰어에게 특정 부분에 대한 집중 혹은 코멘트를 요청하는 경우에도 같이 작성해주면 좋습니다.
- PR을 나눠서 올려야 했는데.. 죄송합니다..ㅠㅠㅠ
- MVVM + Combine 처음 써봐서 이게 맞나..싶어요😂
- VC끼리 데이터 전달(선택한 사진, 선택한 앨범)은 `NotificationCenter` 사용해서 데이터 받은 쪽은 input, 보내는쪽은 output으로 구현했습니다...! (이런 상황에 input, output 맞게 구분한건지 궁금합니다)
- 커스텀 앨범 구현하는게 생각보다 공수가 많이 드네요🥹 실제 디바이스에도 잘 되는지 테스트 해주시면 감사하겠습니당

### 🏞️ 스크린샷
<img src="https://user-images.githubusercontent.com/77915491/217269304-40ee969a-5d93-4b9c-819b-8a011ef0e554.png" width="30%"> <img src="https://user-images.githubusercontent.com/77915491/217269460-2794d274-a79d-49ac-b64d-28b022dd684e.png" width="30%"> <img src="https://user-images.githubusercontent.com/77915491/217269588-993279cd-9333-46b1-9226-e08a39df7b74.png" width="30%">

dudtjs1021ej committed:
✨ [feat] MyFitftyType 추가

dudtjs1021ej committed:
✨ [feat] UploadCodyCoordinator에 MyFitftyType 추가

dudtjs1021ej committed:
✨ [feat] 내 핏프티 수정 UI 구현

dudtjs1021ej committed:
💄 [style] UploadCody textView 스크롤 보이게 수정, height 변경

dudtjs1021ej committed:
✨ [feat] 핏프티 등록 글자수 제한 구현

dudtjs1021ej committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/myFitfty-ios2-149

dudtjs1021ej committed:
🚚 [style] UploadCody에서 MyFitfty로 네이밍 변경

- UploadCodyViewController -> MyFitftyViewController 변경
- UploadCodyCoordinator -> MyFitftyCoordinator 변경
- UploadCodyCoordinatorInterface -> MyFitftyCoordinatorInterface 변경

dudtjs1021ej committed:
✨ [feat] StyleTag 추가

dudtjs1021ej committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/myFitfty-ios2-135

dudtjs1021ej committed:
✨ [feat] AlbumListCoordinator 구현

dudtjs1021ej committed:
💄 [style] 앨범리스트 naivgationBarView 추가

dudtjs1021ej committed:
💄 [style] BarView ChevronButton true 때 레이아웃 수정

dudtjs1021ej committed:
✨ [feat] 앨범 리스트 UI 구현

dudtjs1021ej committed:
✨ [feat] PhotoService로 앨범 불러오기 구현

dudtjs1021ej committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/myFitfty-ios2-135

dudtjs1021ej committed:
✨ [feat] 앨범목록 썸네일 추가

dudtjs1021ej committed:
💄 [style] 앨범 사진 contentMode 변경

dudtjs1021ej committed:
♻️ [refactor] AlbumListViewModel output 수정

dudtjs1021ej committed:
✨ [feat] AlbumViewModel 구현

dudtjs1021ej committed:
✨ [feat] 앨범 목록 선택 구현

dudtjs1021ej committed:
♻️ [refactor] showBottomSheet 자연스럽게 레이아웃 개선

dudtjs1021ej committed:
✨ [feat] 선택한 사진 정보 전달 구현

dudtjs1021ej committed:
✨ [feat] MyFitftySection 구현

dudtjs1021ej committed:
✨ [feat] 선택한 사진 핏프티 등록 화면에 불러오기 구현

dudtjs1021ej committed:
🐛 [fix] 앨범 리스트 썸네일 안나오는 현상 해결

dudtjs1021ej committed:
✨ [feat] 앨범명 영어에서 한글로 변경

dudtjs1021ej committed:
✨ [feat] 사진을 다시 고를 때 터치 영역 확대

dudtjs1021ej committed:
🚚 [style] 줄바꿈 수정

dudtjs1021ej committed:
🐛 [fix] AlbumListCoordinator 메모리 누수 해결

dudtjs1021ej commented on Fitfty/Projects/Common/Sources/BarView.swift:
```diff
@@ -52,16 +54,27 @@ public final class BarView: UIView {
-            cancelButton.widthAnchor.constraint(equalToConstant: 32),
-            cancelButton.heightAnchor.constraint(equalToConstant: 32)
-        ])
+        if isChevronButtonHidden {
+            backgroundStackView.addArrangedSubviews(titleLabel)
+            NSLayoutConstraint.activate([
+                backgroundStackView.centerYAnchor.constraint(equalTo: centerYAnchor),
+                backgroundStackView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 20),
+                cancelButton.centerYAnchor.constraint(equalTo: centerYAnchor),
+                cancelButton.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -20),
+                cancelButton.widthAnchor.constraint(equalToConstant: 32),
+                cancelButton.heightAnchor.constraint(equalToConstant: 32)
+            ])
+        } else {
+            backgroundStackView.addArrangedSubviews(titleLabel, chevronButton)
+            NSLayoutConstraint.activate([
+                backgroundStackView.centerYAnchor.constraint(equalTo: centerYAnchor),
```
앨범쪽 BarView만 피그마 디자인이 수정되어서 
isChevronButtonHidden에 따라 레이아웃을 다르게 잡아주었습니다!

dudtjs1021ej commented on Fitfty/Projects/Common/Sources/FitftyButton.swift:
```diff
@@ -44,6 +44,10 @@ final public class FitftyButton: UIButton {
         self.style = style
     }
     
+    public func getStyle() -> FitftyButtonStyle {
+        return style
+    }
+    
```
버튼이 enable상태일때만 사진 등록을 해야해서 getStyle을 따로 만들었는데 괜찮나여...?

dudtjs1021ej commented on Fitfty/Projects/Common/Sources/StyleTag.swift:
```diff
@@ -7,15 +7,49 @@
 
 public enum StyleTag {
-    case formal
+    case minimal
+    case modern
     case casual
+    case street
+    case lovely
+    case hip
+    case luxury
```
스타일태그 추가했습니당!

leeari95 replied:
오우! 저도 만드려고 했는데, 잘쓸께요 썬~ 🙏🏻

dudtjs1021ej commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewControllers/MyFitftyViewController.swift:
```diff
@@ -88,8 +104,65 @@ final public class UploadCodyViewController: UIViewController {
+        coordinator.dismiss()
+    }
+    
+    @objc func scrollToBottom() {
+        guard collectionView.numberOfSections > 0 else {
+            return
+        }
+        let indexPath = IndexPath(
+            item: collectionView.numberOfItems(inSection: collectionView.numberOfSections - 1) - 1,
+            section: collectionView.numberOfSections - 1
+        )
+        collectionView.scrollToItem(at: indexPath, at: .bottom, animated: true)
+    }
```
텍스트 입력할 때 키보드에 textView가 가려져서 스크롤을 밑으로 내려주었어요!

leeari95 replied:
오... 너무 좋네요 👍👍👍

dudtjs1021ej commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewModel/Album/AlbumListViewModel.swift:
```diff
@@ -0,0 +1,61 @@
+    
+    func selectAlbum(index: Int) {
+        let albumInfo = albums[index]
+        NotificationCenter.default.post(name: .selectAlbum, object: albumInfo)
+    }
```
선택한 앨범을 AlbumVC에 보내줘야해서 NotificationCenter로 보내줬습니다..!

leeari95 commented on Fitfty/Projects/Coordinator/Sources/Main/AlbumListCoordinator.swift:
```diff
@@ -0,0 +1,50 @@
+
+extension AlbumListCoordinator: AlbumListCoordinatorInterface {
+    
+    func dismiss() {
+        navigationController.dismiss(animated: true)
+        finishDelegate?.coordinatorDidFinish(childCoordinator: self)
+    }
```
제가 앱을 실행해서 앨범 리스트를 X 버튼을 통해 닫으니까, 화면이 스르륵 내려가야할 것 같은데...
화면에서 갑자기 뿅!! 하고 사라지는 것 같아요. 
dismiss 메소드의 completion을 활용해보면 어떨까요?

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/37985d933682510378b7628dda9239eb4d650e32
메모리 누수 신경쓰다가 애니메이션 체크를 못했네여..😂 감사합니다 아리!!
```swift
func dismiss() {
        navigationController.dismiss(animated: true) {
            self.finishDelegate?.coordinatorDidFinish(childCoordinator: self)
        }
    }
```

이렇게 개선해보았어요!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/PhotoService.swift:
```diff
@@ -0,0 +1,204 @@
+        PHPhotoLibrary.shared().unregisterChangeObserver(self)
+    }
+    
+    // 앨범들 가져오기
+    func getAlbums(completion: @escaping ([AlbumInfo]) -> Void) {
+        var allAlbums = [AlbumInfo]()
+        defer {
+            completion(allAlbums)
+        }
+        // 스마트 앨범 가져오기 (ex 즐겨찾는 항목)
```
네이밍으로 알아볼 수 있어서
여기 주석은 없어도 괜찮을 것 같아요. ☺️

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/da4ed00402074a76f1beedafd768e82299d8cc49
주석 삭제했습니다!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/PhotoService.swift:
```diff
@@ -0,0 +1,204 @@
+                pointer.pointee = true
+                return
+            }
+            if smartAlbum.estimatedAssetCount == NSNotFound {
```
이 if문은 별도의 else가 없는 것 같은데, guard는 어떨까요?
그럼 코드 depth가 더 나아질 것 같아요!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/da4ed00402074a76f1beedafd768e82299d8cc49
guard문으로 개선해봤어요!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/PhotoService.swift:
```diff
@@ -0,0 +1,204 @@
+                fetchOptions.predicate = Const.predicate
+                fetchOptions.sortDescriptors = Const.sortDescriptors
+                let smartAlbums = PHAsset.fetchAssets(in: smartAlbum, options: fetchOptions)
+                guard 0 < smartAlbums.count else {
+                    return
+                }
+                if smartAlbums.count > 0 {
```
이렇게 두번씩 체크하는 이유가 궁금합니다! 🙋🏻‍♀️

dudtjs1021ej replied:
앗.. 중복코드네요..ㅎㅎㅎ 수정하겠습니다!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/da4ed00402074a76f1beedafd768e82299d8cc49
수정완료!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/PhotoService.swift:
```diff
@@ -0,0 +1,204 @@
+        guard 0 < smartAlbums.count else {
+            return allAlbums
+        }
+        smartAlbums.enumerateObjects { smartAlbum, index, pointer in
+            guard index <= smartAlbums.count - 1 else {
+                pointer.pointee = true
+                return
+            }
+            if smartAlbum.estimatedAssetCount == NSNotFound {
+                let fetchOptions = PHFetchOptions()
+                fetchOptions.predicate = Const.predicate
+                fetchOptions.sortDescriptors = Const.sortDescriptors
+                let smartAlbums = PHAsset.fetchAssets(in: smartAlbum, options: fetchOptions)
+                if smartAlbums.count > 0 {
+                    allAlbums.append(
+                        .init(
+                            identifier: smartAlbum.localIdentifier,
+                            name: self.englishToKorean(title: smartAlbum.localizedTitle),
+                            photoCount: smartAlbums.count,
+                            album: smartAlbums,
+                            thumbnailImage: self.assetToImage(asset: smartAlbums[0])
+                        )
+                    )
+                }
+            }
+        }
```
이 부분이 위에 코드랑 반복되는 것 같아요. 메소드로 분리해볼 수 있을까요? 💪

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/da4ed00402074a76f1beedafd768e82299d8cc49
`smartAlbumsToAlbumInfos` 메소드 만들었습니당(네이밍 괜찮나여...?ㅎ)

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/PhotoService.swift:
```diff
@@ -0,0 +1,204 @@
+    func getPHAssets(album: PHFetchResult<PHAsset>) -> [PHAsset] {
+        var phAssets = [PHAsset]()
+        album.enumerateObjects { asset, index, stopPointer in
+            guard index <= album.count - 1 else {
+                stopPointer.pointee = true
+                return
+            }
+            phAssets.append(asset)
```
이 경우엔 if문이 더 가독성이 좋아질 것 같은데, 어떻게 생각하시나요?

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/da4ed00402074a76f1beedafd768e82299d8cc49
else문에서 
`stopPointer.pointee = true` 를 처리해줬어요!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/PhotoService.swift:
```diff
@@ -0,0 +1,204 @@
+    
+}
+
+// 사진 접근 권한: 선택된 사진
```
이 부분도 제거해도 괜찮을 것 같아요 ㅎㅎ

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/259a2c5a238e4e2ccced78345aa101a97dc10acb
제거완료👍

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewControllers/MyFitftyViewController.swift:
```diff
@@ -9,35 +9,29 @@
     private var styleTagItems : [(styleTag: StyleTag, isSelected: Bool)] = [
-        (.formal, true),
-        (.casual, false)
+        (.minimal, false),
+        (.modern, false),
+        (.casual, false),
+        (.street, false),
+        (.lovely, false),
+        (.hip, false),
+        (.luxury, false)
     ]
+    
     private var weatherTagItems: [(weatherTag: WeatherTag, isSelected: Bool)] = [
-        (.coldWaveWeather, true),
+        (.coldWaveWeather, false),
         (.coldWeather, false),
         (.chillyWeather, false),
         (.warmWeather, false),
```
음.. 이 부분은 ViewModel이 가지고 있는 게 적절하지 않을까요?
뭔가 뷰에 보여줄 데이터 같다는 생각이 들어서요.
지금은 그런 로직이 안보이는 것 같은데, 추후 실제 게시물을 수정할 때도, 이 모델들을 초기화해주어야 할 수도 있구요.
그리고 뷰모델도 수정인지, 새등록인지 MyFitftyType을 가지고 있어야할 것 같아요.

dudtjs1021ej replied:
tagItems들을 viewModel에서 가지고 있으면 
```swift
// MyFitftyViewController - setUpDataSource()

 case .weatherTag:
                    let cell = collectionView.dequeueReusableCell(WeatherTagCell.self, for: indexPath)
                    cell?.setUp(
                        weahterTag: self.weatherTagItems[indexPath.item].weatherTag,
                        isSelected: self.weatherTagItems[indexPath.item].isSelected
                    )
                    return cell ?? UICollectionViewCell()
```

요기 cell setUp 처리는 어떻게 해야 할까요?? ViewModelState를 하나 만들고 태그를 send해주면 될까여..?
게시물 수정할때 초기화는 해당 게시물의 선택 되었던 태그들 불러오는거 말하시는거 맞나용?

leeari95 replied:
넵 맞습니당.
어짜피 viewModel에서 sections를 send로 보낼 때마다, snapshot이 동작할탠데,
MyFitftyCellModel 타입을 별도로 구현해두고... 케이스별로 다른 타입을 가질 수 있도록 하면 어떨까요?

예를들어 태그라는 모델을 만들어서 name과, isSelected을 가지고 있고,
그 모델을 통해서 setUp 메소드 파라미터로 전달해주면 좋을 거 같아요.

예시는... MainFeed에 `MainFeedSection.swift` 파일의 MainCellModel 타입을 참고해주세요 !

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/a08cd7accaf574d5d30711b718e6b2c58a549f59
tagItems 뷰모델에서 관리하게 개선해봤어용

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewControllers/MyFitftyViewController.swift:
```diff
@@ -88,8 +104,65 @@ final public class UploadCodyViewController: UIViewController {
+        viewModel.state.compactMap { $0 }
+            .sinkOnMainThread(receiveValue: { [weak self] state in
+                switch state {
+                case .sections(let sections):
+                    self?.applySnapshot(sections)
+                case .reload(let sections, let image):
+                    self?.selectedImage = image
+                    self?.applySnapshot(sections)
```
음.. 이미 sections라는 state값이 있는데, reload에도 sections를 추가해준 이유, 썬의 생각이 궁금하네요.
저라면 image를 바꾸는 이벤트, sections를 업데이트하는 이벤트 각각 보내볼 것 같다는 생각이 들어서요!

dudtjs1021ej replied:
imge를 받아오고 다시 applySnapshot을 해줘야해서 섹션도 추가했는데
currentState.send를 sections 상태타입도 보내면 되는군요..! 수정해보겠습니당

dudtjs1021ej replied:
```swift
currentState.send(.reload(phAssetInfo.image))
currentState.send(.sections([
                MyFitftySection(sectionKind: .content, items: [UUID()]),
                MyFitftySection(sectionKind: .weatherTag, items: Array(0..<5).map { _ in UUID() }),
                MyFitftySection(sectionKind: .styleTag, items: Array(0..<7).map { _ in UUID() })
]))
```

이미지 선택했을 때 reload, sections 두 번 보내게 개선했어요
아 지금 보니 reload 네이밍이 좀 이상한 것 같은데 괜찮나요..?ㅋㅋㅋ

leeari95 replied:
음... 코디사진은 어떤가용?

dudtjs1021ej replied:
오옹 codyImage로 수정해볼게요!

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/9e0f9ce59b8309770481d5a43958e31b65a345b5
수정했어용

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewModel/Album/AlbumListViewModel.swift:
```diff
@@ -0,0 +1,61 @@
+protocol AlbumListViewModelOutput {
+    
+    var output: AlbumListViewModelOutput { get }
+    func selectAlbum(index: Int)
```
Input에 있어야 할 메소드 같아요!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewModel/Album/AlbumViewModel.swift:
```diff
@@ -0,0 +1,104 @@
+protocol AlbumViewModelOutput {
+    
+    var output: AlbumViewModelOutput { get }
+    func selectImage(index: Int)
```
이 부분도요!

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewModel/MyFitfty/MyFitftyViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+    
+}
+
+protocol MyFitftyViewModelOutput {
+    
+    var output: MyFitftyViewModelOutput { get }
+    
+}
```
output이 존재하지 않는데, 별도로 정의해주신 이유가 궁금하네요. 🤔

dudtjs1021ej replied:
나중에 코디 등록 구현할 때 output이 필요할 것 같아 미리 정의만 해두었는데.. 다시 생각해봐야 할 것 같네여ㅎ
output 코드 삭제하고 다시 올리겠습니당

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/8558f5875564239e134e6baade83b6e8a8fd51e8
지웠어요ㅎㅎ

leeari95 reviewed (changes requested):
안녕하세요 썬~ 기능 구현하시느라 고생 많으셨습니다!!! 👏👏👏
앨범 너무 구현 잘해주신 것 같아요. 일단 시뮬레이터로 테스트 해봤는데, 문제 없는 것 같습니다.


## 저의 의견

> VC끼리 데이터 전달(선택한 사진, 선택한 앨범)은 NotificationCenter 사용해서 데이터 받은 쪽은 input, 보내는쪽은 output으로 구현했습니다...! (이런 상황에 input, output 맞게 구분한건지 궁금합니다)

아.. 이 부분 코드상에서는 뭔가 input 역할을 하는 메소드 같았는데, NotificationCenter 때문이였군요.
혹시 `delegate 패턴`은 고려해보시지 않으셨을까요?

코디네이터로 앨범목록을 띄워줄탠데, 그때 delegate도 같이 넘겨주어 구현해볼 수도 있을 것 같아요.
NotificationCenter은 아무래도 코드 흐름을 파악하기 어렵다고 해야할까요?
지금같이 코드양이 적을 때는 모르겠는데, 추후 유지보수할 코드 양이 많아진다면... 
코드 파악하는데 시간이 걸릴 것 같아서요. 한번 고민해보시고, 가능하다면 delegate 패턴으로 개선해보는 것도 좋을 것 같아요 ㅎㅎ

# 

자세한 사항은 코멘트로 남겨두었으니 확인 부탁드립니다.
혹시 의문이 가거나, 추가로 궁금하신 사항 있으면 언제든 코멘트나 디엠주시면 감사하겠습니다!!
고생 많으셨습니다~ 😊

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewControllers/MyFitftyViewController.swift:
```diff
@@ -57,13 +51,33 @@ final public class UploadCodyViewController: UIViewController {
         return collectionView
     }()
     
+    private lazy var disableRightBarButton: UIBarButtonItem = {
+        let button = UIButton()
+        button.setTitle(myFitftyType.buttonTitle, for: .normal)
+        button.setTitleColor(CommonAsset.Colors.gray03.color, for: .normal)
+        button.titleLabel?.font = FitftyFont.appleSDSemiBold(size: 16).font
+        return UIBarButtonItem(customView: button)
+    }()
+    
+    private lazy var enableRightBarButton: UIBarButtonItem = {
+        let button = UIButton()
+        button.setTitle(myFitftyType.buttonTitle, for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.titleLabel?.font = FitftyFont.appleSDMedium(size: 15).font
+        button.frame = CGRect(x: 0, y: 0, width: 65, height: 37)
+        button.backgroundColor = .black
+        button.layer.cornerRadius = 18
+        button.addTarget(self, action: #selector(didTapUploadButton), for: .touchUpInside)
+        return UIBarButtonItem(customView: button)
+    }()
```
음... 하나의 버튼으로 합치고,
사용자가 등록을 할 수 있는 상태인지, 아닌지를 판별해서 ViewModel에서 이벤트를 줄때마다
버튼의 스타일을 변경하는 건 어떤가요?

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewControllers/AlbumViewController.swift:
```diff
@@ -93,43 +68,61 @@ final public class AlbumViewController: UIViewController {
-                self?.phAssets = phAssets
-            }
+    @objc private func didTapUploadButton(_ sender: Any?) {
+        guard uploadButton.getStyle() == .enabled else {
+            return
+        }
+        if let selectedIndex = selectedIndex {
+            viewModel.output.selectImage(index: selectedIndex)
         }
+        coordinator.dismiss()
```
음... 버튼의 상태를 바라보고 있는 방법보다는, ViewModel 단에서
사용자가 업로드를 할 수 있는 상태인지를 output으로 내보내주면 어떨까요?
그럼 구현해둔 getStyle() 메소드도 필요없어질 것 같아요.

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/90057bbdfcfe4bdee1a615c97b80925ce08800ba
`ViewModelState` 에 `completed(Bool)` 을 추가하고
input으로 `didTapUpload(isSelected: Bool)` 을 감지해서 사진이 선택되었을때만 dismiss되게 구현했는데.. 괜찮나요??

leeari95 replied:
제가 놓쳤던 부분이 있었네용...?
이제보니 selectedIndex도 ViewController가 아니라 ViewModel이 가지고 있는게 맞는 것 같습니다!!

leeari95 replied:
202번 줄에 selectedIndex를 할당해주고 있는데, 이걸 input으로 보내면 어떨까요?

leeari95 replied:
그리고...
state에 값이 들어왔을 때, viewModel.output으로 selectImage를 보내는 것도 약간 이상한 것 같네요....

왜 이상하냐면, View는 ViewModel에서 받아온 데이터를 화면에 표시하기만 해야하는데... 
추가 동작을 더 하는 느낌이에요.

제가 생각하는[?] MVVM 중 ViewController의 역할은
사용자에게 받은 이벤트를 뷰모델로 쏴주고(`input`), 뷰모델에게 받은 데이터를 표시(`state`)해준다고 생각해요.
근데 현재 저기 로직은 뷰를 닫기(coordinator.dismiss)만 해야할 것 같은데, 뷰모델로 또 이벤트를 보내고 있네용.

이 부분은 다시한번 고민해보시면 좋을 것 같습니다!

completed에서는 코디네이터만 호출되는 것이 적절할 것 같아요.

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/8db8487dbc37f52abcc8d5024fb5bbbcebdc41c1
AlbumViewModel에서 selectedIndex를 관리하도록 개선했어요!

<br>
<br>

https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/b5b1196a3d9165d1b225d5491efa3799bb20f101
>completed에서는 코디네이터만 호출되는 것이 적절할 것 같아요.

이 부분 고민해봤는데..ViewModelState에 sendImage를 추가해서

```swift
 case .completed(let completed):
          guard completed else {
              return
          }
          self?.coordinator.dismiss()
case .sendImage:
          self?.viewModel.output.selectImage()

```

이렇게 개선해봤는데 제가 맞게 이해했나요?🥲

leeari95 replied:
흠... 이 부분은 같이 음성으로 이야기 나눠보면 좋을 것 같습니당!! 글로 하려니까 한계가 있네요 ㅋㅋㅋ🥹

dudtjs1021ej replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/9beef418feed705e27a4a1187eb12f6ee624133d
sendImage state지우고 didTapUpload()에서 바로 이미지 보내게 개선했어요!!
감사합니다 아리👍

leeari95 commented on Fitfty/Projects/MainFeed/Sources/MyFitfty/ViewControllers/MyFitftyViewController.swift:
```diff
@@ -57,13 +51,33 @@ final public class UploadCodyViewController: UIViewController {
         return collectionView
     }()
     
+    private lazy var disableRightBarButton: UIBarButtonItem = {
+        let button = UIButton()
+        button.setTitle(myFitftyType.buttonTitle, for: .normal)
+        button.setTitleColor(CommonAsset.Colors.gray03.color, for: .normal)
+        button.titleLabel?.font = FitftyFont.appleSDSemiBold(size: 16).font
+        return UIBarButtonItem(customView: button)
+    }()
+    
+    private lazy var enableRightBarButton: UIBarButtonItem = {
+        let button = UIButton()
+        button.setTitle(myFitftyType.buttonTitle, for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.titleLabel?.font = FitftyFont.appleSDMedium(size: 15).font
+        button.frame = CGRect(x: 0, y: 0, width: 65, height: 37)
+        button.backgroundColor = .black
+        button.layer.cornerRadius = 18
+        button.addTarget(self, action: #selector(didTapUploadButton), for: .touchUpInside)
+        return UIBarButtonItem(customView: button)
+    }()
```
버튼 이름 때문에 오해했네요 ㅋㅋㅋ
enable, disable에 따라 표시해주는 버튼인 줄 알았어요ㅋㅋㅋ😂
등록, 수정 버튼 같은데, 네이밍을 다시한번 고민해보면 좋을 것 같아요!

dudtjs1021ej replied:
enable, disable에 따라 표시해주는 버튼이 맞아요!
처음 VC가 로드될때 수정 화면은 rightBarButton이 활성화 되어있고 등록 화면은 비활성화 되어있어서 setNavigtaionBar()에서 그렇게 처리해주었습니다..!
등록화면에서 모든 태그를 선택하고 이미지랑 텍스트 입력이 되면 enable버튼으로 바꾸는 로직 추가할 예정입니당

dudtjs1021ej committed:
♻️ [refactor] 앨범 목록 dismiss 자연스럽게 개선

dudtjs1021ej committed:
♻️ [refactor] PhotoService 코드 개선

- if문, guard문 개선
- smartAlbumsToAlbumInfos메소드 추가
- 주석 제거

dudtjs1021ej committed:
🚚 [style] MyFitftyViewModel 불필요한 output 관련 코드 삭제

dudtjs1021ej committed:
♻️ [refactor] MyFitftyViewModel reload 타입 개선

dudtjs1021ej committed:
♻️ [refactor] 사진 업로드 이벤트 개선

- FitftyButton getStyle() 제거
- AlbumViewModelInput didTapUpload() 추가
- AlbumViewModel state completed 추가

dudtjs1021ej committed:
💡 [comment] 불필요한 주석 제거

dudtjs1021ej committed:
♻️ [refactor] AlbumViewModel이 selectedIndex를 관리하게 개선

dudtjs1021ej committed:
♻️ [refactor] reload에서 codyImage로 명확하지 않은 네이밍 변경

dudtjs1021ej committed:
♻️ [refactor] styleTagItems과 weatherTagItems MyFitftyViewModel에서 관리하게 개선

dudtjs1021ej committed:
✨ [feat] 사진을 선택해도 텍스트가 유지되게 구현

dudtjs1021ej committed:
♻️ [refactor] AlbumViewModel ViewModelState sendImage 추가

dudtjs1021ej committed:
♻️ [refactor] 사진 선택 후 표시될 때 애니메이션 자연스럽게 개선

dudtjs1021ej reviewed (commented):
@leeari95 아리 이번에도 꼼꼼한 리뷰 감사합니다!! 
delegate 패턴은 시간 부족으로... 여유 생기면 리팩토링 해볼게요🥹
<br>

https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/29/commits/ee0ba1e4b23a76221614816651e033739956d141

추가로 사진을 선택하면 텍스트가 사라지는 현상을 발견해서...
delegate 패턴 사용해서 입력한 text를 뷰모델에서 처리하게끔 구현해봤어용

dudtjs1021ej committed:
♻️ [refactor] AlbumViewModel의 didTapUpload()에서 선택한 이미지 보내게 개선
